shared/src/chain_state.rs

```diff
@@ -0,0 +1,357 @@
+use crate::tx_proposal_table::TxProposalTable;
+use crate::txo_set::TxoSet;
+use crate::txo_set::TxoSetDiff;
+use ckb_core::block::Block;
+use ckb_core::cell::{CellProvider, CellStatus, ResolvedTransaction};
+use ckb_core::header::{BlockNumber, Header};
+use ckb_core::transaction::{OutPoint, ProposalShortId, Transaction};
+use ckb_core::Cycle;
+use ckb_verification::{TransactionError, TransactionVerifier};
+use fnv::FnvHashSet;
+use log::error;
+use lru_cache::LruCache;
+use numext_fixed_hash::H256;
+use numext_fixed_uint::U256;
+use std::cell::{Ref, RefCell};
+use std::sync::Arc;
+
+#[derive(Debug, Clone)]
+pub struct ChainState<CI> {
+    store: Arc<CI>,
+    tip_header: Header,
+    total_difficulty: U256,
+    txo_set: TxoSet,
+    proposal_ids: TxProposalTable,
+    // interior mutability for immutable borrow proposal_ids
+    tx_pool: RefCell<TxPool>,
+    txs_verify_cache: RefCell<LruCache<H256, Cycle>>,
+}
+
+impl<CI: ChainIndex> ChainState<CI> {
+    pub fn new(
+        store: &Arc<CI>,
+        tip_header: Header,
+        total_difficulty: U256,
+        txo_set: TxoSet,
+        proposal_ids: TxProposalTable,
+        tx_pool: TxPool,
+        txs_verify_cache: LruCache<H256, Cycle>,
+    ) -> Self {
+        ChainState {
+            store: Arc::clone(store),
+            tip_header,
+            total_difficulty,
+            txo_set,
+            proposal_ids,
+            tx_pool: RefCell::new(tx_pool),
+            txs_verify_cache: RefCell::new(txs_verify_cache),
+        }
+    }
+
+    pub fn tip_number(&self) -> BlockNumber {
+        self.tip_header.number()
+    }
+
+    pub fn tip_hash(&self) -> H256 {
+        self.tip_header.hash()
+    }
+
+    pub fn total_difficulty(&self) -> &U256 {
+        &self.total_difficulty
+    }
+
+    pub fn tip_header(&self) -> &Header {
+        &self.tip_header
+    }
+
+    pub fn txo_set(&self) -> &TxoSet {
+        &self.txo_set
+    }
+
+    pub fn is_spent(&self, o: &OutPoint) -> Option<bool> {
+        self.txo_set.is_spent(o)
+    }
+
+    pub fn contains_proposal_id(&self, id: &ProposalShortId) -> bool {
+        self.proposal_ids.contains(id)
+    }
+
+    pub fn update_proposal_ids(&mut self, block: &Block) {
+        self.proposal_ids
+            .update_or_insert(block.header().number(), block.union_proposal_ids())
+    }
+
+    pub fn get_proposal_ids_iter(&self) -> impl Iterator<Item = &ProposalShortId> {
+        self.proposal_ids.get_ids_iter()
+    }
+
+    pub fn reconstruct_proposal_ids(&mut self, number: BlockNumber) -> Vec<ProposalShortId> {
+        self.proposal_ids.reconstruct(number)
+    }
+
+    pub fn update_tip(&mut self, header: Header, total_difficulty: U256, txo_diff: TxoSetDiff) {
+        self.tip_header = header;
+        self.total_difficulty = total_difficulty;
+        self.txo_set.update(txo_diff);
+    }
+
+    pub fn add_tx_to_pool(&self, tx: Transaction, max_cycles: Cycle) -> Result<(), PoolError> {
```

Comment:
if tx is not proposed, it will push to the pending pool without verify immediately.
It not possible to get actual cycles here, we can improve traces to record verify error and cycles,
but not this PR.